### PR TITLE
feat: constraints api integration

### PR DESCRIPTION
**Pull request by namn-grg**

### CODE REVIEW

File: crates/api/src/builder/api.rs [hunk 1]
```diff
@@ -378,6 +370,225 @@ where
...
+    /// This function extends the `submit_block` functionality to also handle inclusion proofs:
+    /// 1. Receives the request and decodes the payload into a `SignedBidSubmission` object.
+    /// 2. Validates the builder and checks against the next proposer duty.
+    /// 3. Verifies the signature of the payload.
+    /// 4. Fetches and handles inclusion proofs.
+    /// 5. Runs further validations against the auctioneer.
+    /// 6. Simulates the block to validate the payment.
+    /// 7. Saves the bid and proofs to the auctioneer and db.
+    ///
+    /// Implements this API: <https://flashbots.github.io/relay-specs/#/Builder/submitBlockWithProofs>
```
Comment (thedevbirb) [inline crates/api/src/builder/api.rs:384]: This link should point to our docs

File: crates/api/src/builder/api.rs [hunk 2]
```diff
@@ -378,6 +370,225 @@ where
...
+        let (payload, is_cancellations_enabled) = 
+            decode_payload(req, &mut trace, &request_id).await?;
+        let block_hash = payload.message().block_hash.clone();
+
+        // Verify that we have a validator connected for this slot
+        if next_duty.is_none() {
+            warn!(request_id = %request_id, "could not find slot duty");
+            return Err(BuilderApiError::ProposerDutyNotFound);
+        }
+        let next_duty = next_duty.unwrap();
```
Comment (thedevbirb) [inline crates/api/src/builder/api.rs:410]: This is more idiomatic:
```suggestion
        let Some(next_duty) = next_duty else {
            warn!(request_id = %request_id, "could not find slot duty");
            return Err(BuilderApiError::ProposerDutyNotFound);
        }
```

**Review by thedevbirb — COMMENTED** [review]
Good job! Just two little nits I've noticed so far

File: crates/api/src/builder/api.rs (hunk 2 above)
Comment (namn-grg) [reply crates/api/src/builder/api.rs:410]: Thanks for the suggestion I agree, I copied the already written code of `submit_block` where they did it this way. I am trying to not make extra changes and keep the coding style same.

File: crates/api/src/builder/api.rs (hunk 1 above)
Comment (namn-grg) [reply crates/api/src/builder/api.rs]: Thanks for pointing out! Done

File: crates/utils/src/signing.rs [hunk 3]
```diff
@@ -9,6 +9,18 @@ use ethereum_consensus::{
     Error, Fork,
 };
 
+pub fn verify_constraints_signature<T: Merkleized>(
+    message: &mut T,
+    signature: &BlsSignature,
+    public_key: &BlsPublicKey,
+    context: &Context,
```
Comment (thedevbirb) [inline crates/utils/src/signing.rs:16]: As we said `context` can be taken from the `chain_info` that is available in the api.

**Review by thedevbirb — COMMENTED** [review]
Good work! Made a first review. Will do another round more detailed at a later time

File: crates/api/src/builder/api.rs [hunk 4]
```diff
@@ -1133,6 +1358,17 @@ where
...
     }
 
+    async fn verify_inclusion_proof(
+        &self,
+        payload: &SignedBidSubmission,
+        constraints: &[ConstraintsWithProofData],
+    ) -> Result<(), BuilderApiError> {
+        let proofs = payload.proofs().expect("proofs not found");
+        // TODO: Value of root
+        verify_multiproofs(constraints, proofs, FixedBytes::default());
```
[the remark this answers is not in the document]
Comment (namn-grg) [reply crates/api/src/builder/api.rs]: Solved this now, was not sure sure of the root value

File: crates/api/src/router.rs [hunk 5]
```diff
@@ -129,10 +119,24 @@ pub fn build_router(
...
             }
             Route::ValidatorRegistration => {
-                router = router.route(
-                    &route.path(),
-                    get(DataApiProd::validator_registration),
-                );
+                router = router.route(&route.path(), get(DataApiProd::validator_registration));
+            }
+            // TODO: Decide where to keep this route. I think it makes sense to store the constraints in the redis cache.
+            // As it is a temporary (hot) storage. `ConstraintsApiProd` can be a good place to keep this route.
```
[the remark this answers is not in the document]
Comment (namn-grg) [reply crates/api/src/router.rs:125]: Will do this in a stacked PR!

File: crates/common/src/proofs.rs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+//                         <DenebSpec as EthSpec>::MaxBytesPerTransaction,
+//                     >::from(tx.to_vec()));
+
+//                 Ok((tx_hash, tx_root))
+//             })
+//             .collect::<Result<Vec<_>, Eip2718Error>>()?;
+
+//         Ok(Self { message: value, proof_data: transactions })
+//     }
+// }
```
[the remark this answers is not in the document]
Comment (namn-grg) [reply crates/common/src/proofs.rs]: Thanks!

File: crates/common/src/proofs.rs [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    // Get all the leaves from the saved constraints
+    let mut leaves = Vec::with_capacity(proofs.total_leaves());
+
+    // NOTE: Get the leaves from the constraints cache by matching the saved hashes. We need the leaves
+    // in order to verify the multiproof.
+    for hash in proofs.transaction_hashes.iter() {
+        let mut found = false;
+        for constraint in constraints {
+            for (saved_hash, leaf) in &constraint.proof_data {
+                if **saved_hash == ****hash {
```
[the remark this answers is not in the document]
Comment (namn-grg) [reply crates/common/src/proofs.rs]: solved this by taking slice value, should be the same thing (?)

**Review by merklefruit — COMMENTED** [review]
will need to see this tested before I can do a proper review, the changes are quite big here - and I don't know the codebase well yet

**Pull request merged**